mm2src/coins/qrc20.rs

```diff
@@ -119,11 +202,15 @@ impl AsRef<UtxoCoinFields> for Qrc20Coin {
     fn as_ref(&self) -> &UtxoCoinFields { &self.utxo }
 }
 
-enum RpcContractCallType {
+impl qtum::QtumBasedCoin for Qrc20Coin {}
+
+pub enum RpcContractCallType {
```

Comment:
I feel concerned a bit in regards to this. Mixing different contract calls here can potentially make this enum growing indefinitely. We might add more different token types in the future and have more functions in the etomic swap smart contract, so I consider the probability of such an event as high. Could you please refactor it separating different smart contract call types to different enums? You can keep `RpcContractCallType`, it will probably look as follows:
```
RpcContractCallType {
    Erc20(ErcContractCallType),
    EtomicSwap(EtomicSwapContractCallType),
    ...,
}
```

Reply:
I think it may be the reason for a lot of boilerplate code. We have `RpcContractCallType` (smart contract functions that don't change the state) and `ContractCallType` (smart contract functions that change the state). Both of them contain both E/QRC20 functions and Etomic Swap functions. As a result, we will have 4 more enums.

Reply:
Do we need to have 2 different types in this case? The only difference I see is ContractCallType has the `from_script_pubkey` method, which can be easily added to the `RpcContractCallType`. So these structs look the same to me.

Reply:
`RpcContactCallType` is passed into `rpc_contract_call` method, and it's not clearer for me if the function will not support all types (like `Transfer`, `Erc20Payment` etc).

Reply:
I mean that if we have 2 enums, which look similar, then we might want to unite them into a single one. I think it is not a big problem if we can't extract the functions, which don't change state, from script pubkey. Alternatively, it's worth considering to rename the enums to something that reflects the actual meaning (functions that change the state and functions that don't).

Reply:
`ContractCallType` is used also as argument in [find_swap_contract_call_with_swap_id](https://github.com/KomodoPlatform/atomicDEX-API/blob/qrc20-native-mode/mm2src/coins/qrc20/swap.rs#L887) function, `ContractCallType` is checked in [is_receiver_contract](https://github.com/KomodoPlatform/atomicDEX-API/blob/qrc20-native-mode/mm2src/coins/qrc20/history.rs#L694), [is_sender_contract](https://github.com/KomodoPlatform/atomicDEX-API/blob/qrc20-native-mode/mm2src/coins/qrc20/history.rs#L670). In these places it might be strange to pass/check the immutable contract call funcitons.

I tried to dive into the naming... As I found [here](https://ethereum.stackexchange.com/questions/28504/is-there-any-profit-from-using-pure-and-view-functions-modifiers) functions that don't change the state are indicated by`view` keyword.
What about `ViewContractCallType` and `PayableContractCallType`? Or `ContractCallType` and `MutContractCallType`.

Reply:
`ViewContractCallType` and `MutContractCallType` looks the best to me. `PayableContractCallType` would be incorrect, because the payable function in solidity means that it can accept the ETH. But smart contract state can be mutated in a non-payable function: e.g. Q/ERC20 `transfer` function is not payable, but it changes the state. Could you also leave a doc comment for these enums, please? To ensure that it is as clear as possible :slightly_smiling_face:

Reply:
Sure, thank you :)

Reply:
Done